### PR TITLE
Add `ones_matrix`

### PR DESCRIPTION
**Pull request by YueRen**
I often need to create a matrix of ones (e.g., to prepend a row of ones when homogenizing matrices in polyhedral geometry), and the added `AbstractAlgebra.ones_matrix` would be the to `Base.ones` what `AbstractAlgebra.zero_matrix` is to `Base.zeros`.

I also think `ones_matrix` is a better name than `one_matrix`, as the latter can be confused with the identity matrix (unlike `zero_matrix` which is the zero element of its ring).

### CODE REVIEW

File: src/Matrix.jl [hunk 1]
```diff
@@ -6503,6 +6503,28 @@ function zero_matrix(R::NCRing, r::Int, c::Int)
...
+#   Ones matrix
+#
+################################################################################
+
+@doc raw"""
+    ones_matrix(R::Ring, r::Int, c::Int)
+
+Return the $r \times c$ ones matrix over $R$.
+"""
+function zero_matrix(R::NCRing, r::Int, c::Int)
```
Comment (joschmitt) [inline src/Matrix.jl:6517]:
```suggestion
function ones_matrix(R::NCRing, r::Int, c::Int)
```

**Comment by YueRen** [comment]
Some of the tests are failing, is it related to the changes that I have made?  The logs claim `test/generic/Matrix-test.jl:111` which is code that I haven't touched.

**Comment by YueRen** [comment]
The tests are failing, but the log does not really show where and how.  Is it possible to see more detail than what is being shown here:
https://github.com/Nemocas/AbstractAlgebra.jl/actions/runs/9429639772/job/26012204170?pr=1725 ?

**Review by fingolfin — CHANGES_REQUESTED** [review]
I support the addition of this method in general.

But in the current form it will only ever produce `Matrix` instances. We really also need a generic default implementation that works for e.g. `ZZMatrix` (when Nemo is loaded), etc. 

I am happy to explain how to do that later, but I now have to quickly run else I won't get any lunch today

**Comment by YueRen** [comment]
@fingolfin Good point, does that mean that I need to add similar functions to `Nemo/src/flint/fmpz_mat.jl` and `Nemo/src/flint/fmpq_mat.jl`?  That should be a problem, I'd be happy to do that.

File: src/Matrix.jl [hunk 2]
```diff
@@ -6503,6 +6503,28 @@ function zero_matrix(R::NCRing, r::Int, c::Int)
...
+#
+################################################################################
+
+@doc raw"""
+    ones_matrix(R::Ring, r::Int, c::Int)
+
+Return the $r \times c$ ones matrix over $R$.
+"""
+function ones_matrix(R::NCRing, r::Int, c::Int)
+   arr = Matrix{elem_type(R)}(undef, r, c)
```
Comment (thofma) [inline src/Matrix.jl:6518]:
```suggestion
   arr = zero_matrix(R, r, c)
```
and then return `arr` at the end. Not sure if this is what @fingolfin had in mind, but it should work.

Comment (YueRen) [reply src/Matrix.jl:6518]: This should works, but wouldn't an implementation of `ones_matrix` for `QQMatrix` and `ZZMatrix` in Nemo be preferable?

Comment (lgoettgens) [reply src/Matrix.jl:6518]: This change here should be applied in any case. Adding special implementations in Nemo can only be done later

Comment (fingolfin) [reply src/Matrix.jl:6518]: I concur with the solution proposed by @thofma here, it seems to be the best we can do *right now*. And @lgoettgens of course also got it right: we first need a generic implementation here, then we can add specialized ones elsewhere.

All in all there are other function in here which ought to be more generic. I think it is time to revive PR #558 . I am looking into that right now.

But that shouldn't delay this PR. Just maybe also mention the new function in a `@docs` block in `docs/src/matrix.md` near `identity_matrix` or so

**Comment by YueRen** [comment]
I need some help with the following problem.

With the current version:
```
function ones_matrix(R::NCRing, r::Int, c::Int)
   arr = zero_matrix(R, r, c)
   for i in 1:r
      for j in 1:c
         arr[i, j] = one(R)
      end
   end
   z = Generic.MatSpaceElem{elem_type(R)}(R, arr) # ERROR HERE
   return z
end
```

I get the following error:
```
julia> M = ones_matrix(ZZ,2,3)
ERROR: MethodError: no method matching AbstractAlgebra.Generic.MatSpaceElem{…}(::AbstractAlgebra.Integers{…}, ::AbstractAlgebra.Generic.MatSpaceElem{…})

Closest candidates are:
  AbstractAlgebra.Generic.MatSpaceElem{T}(::NCRing, ::Matrix{T}) where T<:NCRingElement
   @ AbstractAlgebra ~/.julia/dev/AbstractAlgebra/src/generic/GenericTypes.jl:1073
  AbstractAlgebra.Generic.MatSpaceElem{T}(::NCRing, ::AbstractMatrix{T}) where T<:NCRingElement
   @ AbstractAlgebra ~/.julia/dev/AbstractAlgebra/src/generic/GenericTypes.jl:1079
  AbstractAlgebra.Generic.MatSpaceElem{T}(::NCRing, ::Int64, ::Int64, ::Vector{T}) where T<:NCRingElement
   @ AbstractAlgebra ~/.julia/dev/AbstractAlgebra/src/generic/GenericTypes.jl:1083
  ...

Stacktrace:
 [1] ones_matrix(R::AbstractAlgebra.Integers{BigInt}, r::Int64, c::Int64)
   @ AbstractAlgebra ~/.julia/dev/AbstractAlgebra/src/Matrix.jl:6524
 [2] top-level scope
   @ REPL[5]:1
Some type information was truncated. Use `show(err)` to see complete types.
```

I tried fixing it with it by changing the code to
```
function ones_matrix(R::NCRing, r::Int, c::Int)
   z = zero_matrix(R, r, c)
   for i in 1:r
      for j in 1:c
         z[i, j] = one(R)
      end
   end
   return z
end
```

But `zero_matrix` does not seem to produce the matrix of the desired type:
```
julia> M = ones_matrix(ZZ,2,3)
[1   1   1]
[1   1   1]

julia> typeof(M) # not ZZMatrix
AbstractAlgebra.Generic.MatSpaceElem{BigInt}

julia> typeof(zero_matrix(ZZ,2,3)) # not ZZMatrix either
AbstractAlgebra.Generic.MatSpaceElem{BigInt}

```

Are we sure that we want AbstractAlgebra to return a `ZZMatrix`?

**Comment by lgoettgens** [comment]
```julia
function ones_matrix(R::NCRing, r::Int, c::Int)
   z = zero_matrix(R, r, c)
   for i in 1:r
      for j in 1:c
         z[i, j] = one(R)
      end
   end
   return z
end
```
looks like a correct change. You will only get it return a ZZMatrix if you additionally load Nemo (AA does not know this type)

**Review by fingolfin — APPROVED** [review]
Looks good to me but could you please also add `ones_matrix` to `docs/src/matrix.md`, say next to the blocks for `identity_matrix`

**Comment by YueRen** [comment]
Sorry, I totally forgot @fingolfin request.  I've now added `ones_matrix` below `identity_matrix` in `docs/src/matrix.md`.  Hope that works, I'm a bit unsure since some functions are merely listed in `docs/src/matrix.md` while others have a full docstring with examples in there.

**Review by fingolfin — APPROVED** [review]

**Pull request merged**